File: SortQuickSort.c
#include <stdio.h>
void swap(int *a, int *b) {
   int temp;
   temp = *a;
   *a = *b;
   *b = temp;
}
int Partition(int a[], int low, int high) {
   int pivot, index, i;
   index = low;
   pivot = high;
   for(i=low; i < high; i++) {
      if(a[i] < a[pivot]) {
         swap(&a[i], &a[index]);
         index++;
      }
   }
   swap(&a[pivot], &a[index]);
   return index;
}
int RandomPivotPartition(int a[], int low, int high) {
   int pvt, n, temp;
   n = rand();
   pvt = low + n%(high-low+1);
   swap(&a[high], &a[pvt]);
   return Partition(a, low, high);
}
int QuickSort(int a[], int low, int high) {
   return 0;
}
int main() {
   int n, i;
   n=7;
   int arr[]={7,4,2,6,3,1,5};
   int high = n-1;
   int low = 0 ;
   int pindex;
   if(low < high) {
      pindex = RandomPivotPartition(arr, low, high);
      QuickSort(arr, low, pindex-1);
      QuickSort(arr, pindex+1, high);
   }
   for (i = 0; i < n; i++)
      printf("%d ", arr[i]);
   return 0;
}
/*function partitionFunc(left, right, pivot)
   leftPointer = left
   rightPointer = right - 1

   while True do
      while A[++leftPointer] < pivot do
         //do-nothing            
      end while
		
      while rightPointer > 0 && A[--rightPointer] > pivot do
         //do-nothing         
      end while
		
      if leftPointer >= rightPointer
         break
      else                
         swap leftPointer,rightPointer
      end if
		
   end while 
	
   swap leftPointer,right
   return leftPointer
	
end function*/

File: Sorting/SortQuickSort.py
import random
import time
def qS(arr,low,high):
    if low<high: # Base Case
        pIndex = partition(arr,low,high) 
        qS(arr,low,pIndex-1) # Left Half
        qS(arr,pIndex+1,high) # Right Half
        
def partition(arr,low,high):
    pivot = arr[low] # First Element as pivot
    i = low 
    j = high
    while i < j:
        # Move i to the right until an element > pivot is found
        while i <= high - 1 and arr[i] <= pivot:
            i += 1
        # Move j to the left until an element <= pivot is found
        while j >= low + 1 and arr[j] > pivot:
            j -= 1
        
        if i < j:
            arr[i], arr[j] = arr[j], arr[i]
            
    # Swap the pivot element (at low) with the element at j
    arr[low], arr[j] = arr[j], arr[low]
    return j
    
arr = [random.randint(0,10) for _ in range(0,1000)]
# print(arr)
curr = time.time()
qS(arr,0,len(arr)-1)
t = time.time() - curr
# print(arr)
print(f'{t*1000:.2f}','ms')

'''
def partition():
    After picking pivot take i++ and j-- on both ends
    From i check if arr[i] > pivot STOP
    From j check if arr[j] < pivot STOP
    SWAP(i,j)
    Repeat until pointers crosses
    SWAP(pivot,i)
'''
'''
Concept: Low and High pointers, no new structures needed

def qS():
    1). Pivot (Pick and place in its right place): 
        1. 1st Element 
        2. Last Element
        3. Median
        4. Random

    2). Smaller on left and Larger on right

    3). Repeat Above Steps
'''